src/kaleidoscope/device/dygma/raise/Focus.cpp

```diff
@@ -0,0 +1,156 @@
+/* -*- mode: c++ -*-
+ * kaleidoscope::device::dygma::Raise -- Kaleidoscope device plugin for Dygma Raise
+ * Copyright (C) 2017-2019  Keyboard.io, Inc
+ * Copyright (C) 2017-2019  Dygma Lab S.L.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifdef ARDUINO_SAMD_RAISE
+
+#include <Kaleidoscope.h>
+#include <Kaleidoscope-FocusSerial.h>
+#include "kaleidoscope/device/dygma/raise/Focus.h"
+
+namespace kaleidoscope {
+namespace device {
+namespace dygma {
+namespace raise {
+
+#ifndef RAISE_FIRMWARE_VERSION
+#define RAISE_FIRMWARE_VERSION "<unknown>"
+#endif
+
+EventHandlerResult Focus::onFocusEvent(const char *command) {
+  if (::Focus.handleHelp(command, PSTR("hardware.version\nhardware.side_power\nhardware.side_ver\nhardware.sled_ver\nhardware.sled_current\nhardware.layout\nhardware.joint\nhardware.keyscan\nhardware.crc_errors\nhardware.firmware\nhardware.flash_left_side\nhardware.flash_right_side\nhardware.verify_left_side\nhardware.verify_right_side")))
+    return EventHandlerResult::OK;
+
+  if (strncmp_P(command, PSTR("hardware."), 9) != 0)
+    return EventHandlerResult::OK;
+
+  if (strcmp_P(command + 9, PSTR("version")) == 0) {
+    ::Focus.send("Dygma Raise");
+    return EventHandlerResult::EVENT_CONSUMED;
+  }
+
+  if (strcmp_P(command + 9, PSTR("firmware")) == 0) {
+    ::Focus.send(RAISE_FIRMWARE_VERSION);
```

Comment:
`RAISE_FIRMWARE_VERSION` is set to `<unknown>`, unless already defined (via compiler flags). The idea is that the build system will fill this out for us, once we get there.